sys/include/net/ieee802154/radio.h

```diff
@@ -0,0 +1,1085 @@
+     * finish the CCA procedure and get the channel status.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     *
+     * @return 0 if request was OK
+     * @return -EAGAIN if the request cannot be performed immediately.
+     * @return negative errno on error
+     */
+    int (*request_cca)(ieee802154_dev_t *dev);
+
+    /**
+     * @brief Confirmation function for @ref ieee802154_radio_ops::request_cca
+     *
+     * This function must be called to finish the CCA procedure.  This
+     * function should be called on @ref IEEE802154_RADIO_CONFIRM_CCA,
+     * If no interrupt is available, this function can be polled.
+     *
+     * @pre @ref call to request_cca was successful.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     *
+     * @return positive number if the channel is clear
+     * @return 0 if the channel is busy
+     * @return -EAGAIN if the CCA procedure hasn't finished.
+     * @return negative errno on error
+     */
+    int (*confirm_cca)(ieee802154_dev_t *dev);
+
+    /**
+     * @brief Get a cap from the radio
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param cap cap to be checked
+     *
+     * @return true if the radio supports the cap
+     * @return false otherwise
+     */
+    bool (*get_cap)(ieee802154_dev_t *dev, ieee802154_rf_caps_t cap);
+
+    /**
+     * @brief Set the threshold for the Energy Detection (first mode of CCA)
+     *
+     * @pre the device is on
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] threshold the threshold in dBm.
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_cca_threshold)(ieee802154_dev_t *dev, int8_t threshold);
+
+    /**
+     * @brief Set CCA mode
+     *
+     * All radios MUST at least implement the first CCA mode (ED Threshold).
+     *
+     * @pre the device is on
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] mode the CCA mode
+     *
+     * @return 0 on success
+     * @return -ENOTSUP if the mode is not supported
+     * @return negative errno on error
+     */
+    int (*set_cca_mode)(ieee802154_dev_t *dev, ieee802154_cca_mode_t mode);
+
+    /**
+     * @brief Set IEEE802.15.4 PHY configuration (channel, TX power)
+     *
+     * This function SHOULD NOT validate the PHY configurations unless
+     * it's specific to the device. The upper layer is responsible of all kind
+     * of validations.
+     *
+     * @pre the device is on
+     * @post the upper layer calls @ref
+     * ieee802154_radio_ops::request_set_trx_state immediately after calling
+     * this function.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] conf the PHY configuration
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*config_phy)(ieee802154_dev_t *dev, ieee802154_phy_conf_t *conf);
```

Comment:
I would say an error should be reported.

Reply:
See https://github.com/RIOT-OS/RIOT/pull/14371/commits/e941563537045434eea9b0ed778121e2b4531a4a

Reply:
Should we also be able to configure channel spacing, chip rate etc here?
A solution could be to extend `ieee802154_phy_conf_t` like so

```C
typedef struct {
    ieee802154_phy_conf_t base;
    uint8_t rate;
    uint8_t chips;
    …
} ieee802154_mr_oqpsk_phy_conf_t;
```

without having to change the API.

Reply:
exactly! That's why I added in the beginning an `ieee802154_phy_config_t` struct, to add members without breaking the API. But never thought about extending it, so that would be really nice!

Reply:
How do we indicate what phy mode the radio should be in?

Reply:
it's assumed that the upper layer sets the PHY configuration, so this information is already there in the MIB of the MAC.

Reply:
What's a MIB?

Reply:
Th MAC Information Base, the list of variables that keep track of the MAC state.

Reply:
Does this yet exist?

Reply:
hmmm we have traces of it in the `netdev_ieee802154_t` struct (channel, page, etc).
But the idea is that this should be handled by the upper layer anyway.

what do you think?

Reply:
Yea we'll need a place to manage channel spacing, center frequency, allowed channels, rates, etc. based on region anyway.

The drivers shouldn't need to worry which part of the world they are used in but just receive configuration parameters to comply with the local regulations from an upper layer that keeps all those gory tables contained in one place (and ideally configurable at compile-time, so we don't need to store the tables for the US, China, etc. in every device - would be nice if that was an option though)